tests/rbd_mirror/test_rbd_mirror_secondary_full.py

```diff
@@ -0,0 +1,122 @@
+
+    Test Case Flow:
+        1. Create a pool on both clusters.
+        2. Create an Image on primary mirror cluster in same pool.
+        3. Configure mirroring (peer bootstrap) between two clusters.
+        4. Enable image mode snapshot based mirroring on the pool respectively.
+        5. Start running IOs on the primary image.
+        6. keep on running IOs till secondary cluster becomes full
+        7. verify after secondary becomes full, mirroring should fail
+"""
+
+
+import json
+
+from tests.rbd.rbd_utils import Rbd
+from tests.rbd_mirror import rbd_mirror_utils as rbdmirror
+from utility.log import Log
+
+log = Log(__name__)
+
+
+def get_total_avail(cluster):
+    """Method to retrieve total available space in cluster"""
+    out = cluster.exec_cmd(cmd="ceph df -f json", output=True)
+    data = json.loads(out)
+    total_avail_bytes = data["stats"]["total_avail_bytes"]
+    total_avail_gb = total_avail_bytes / (1024**3)
+    return int(total_avail_gb)
+
+
+def run(**kw):
+    """Verification of mirroring failure,
+    when secondary cluster storage is full.
+
+    Args:
+        **kw: test data
+
+    Returns:
+        0 - if test case pass
+        1 - if test case fails
+    """
+
+    try:
+        log.info(
+            "Starting CEPH-9507 - Verification of mirroring failure "
+            "when secondary cluster storage is full"
+        )
+        rbd = Rbd(**kw)
+        config = kw.get("config")
+        mode = config.get("mode")
+        mirrormode = config.get("mirrormode")
+        poolname = rbd.random_string() + "_test_pool"
+        imagename = rbd.random_string() + "_test_image"
+        imagespec = poolname + "/" + imagename
+
+        mirror1 = rbdmirror.RbdMirror(
+            kw.get("ceph_cluster_dict").get("ceph-rbd1"), config
+        )
+        mirror2 = rbdmirror.RbdMirror(
+            kw.get("ceph_cluster_dict").get("ceph-rbd2"), config
+        )
+
+        cluster1_size = get_total_avail(mirror1)
+        cluster2_size = get_total_avail(mirror2)
+
+        if cluster2_size > cluster1_size:
+            log.error("secondary cluster more than primary hence test cannot perform")
+            return 1
+
+        image_size = str(cluster2_size) + "GB"
+
+        mirror1.initial_mirror_config(
+            mirror2,
+            poolname=poolname,
+            imagename=imagename,
+            imagesize=image_size,
+            io_total=config.get("io-total", "10G"),
+            mode=mode,
+            mirrormode=mirrormode,
+            peer_mode=config.get("peer_mode", "bootstrap"),
+            rbd_client=config.get("rbd_client", "client.admin"),
+            build=config.get("build"),
+            **kw
+        )
+
+        out, err = mirror1.benchwrite(
+            imagespec=imagespec,
+            io=image_size,
+            long_running=False,
+            all=True,
+            check_ec=False,
+        )
+        log.debug(err)
+
+        # Validation for IO completion
+        if err:
```

Comment:
Are you filling the cluster and also checking for mirroring in the same step? I don't think this is a good approach for this test case.

Afaik what you should do is, first fill the secondary cluster upto the stage where you start getting a HEALTH_ERR saying OSDs are full. 
Now try to create another image in primary and enable mirroring on this image, this should not succeed since ther e is no space left for the mirrored image to be created in secondary.

Instead you are just trying to write IOs to an already mirrored image which is not what this test case - CEPH-9507 is trying to test.

Reply:
Done